# Projekt IAU  
## 1. časť - Prieskumná analýza 
### Vypracovali: Martin Kukučka, Martin Jankuliak

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
import glob

import datetime
from datetime import datetime,date
from datetime import timedelta
from dateutil.relativedelta import relativedelta

import json
import re
import dateutil.parser as parser
import statsmodels.api as sm

import scipy as sc
from pandas import read_csv


### Načítanie dát

In [3]:
df_other_train = read_csv("data/other_train.csv")
df_personal_train = read_csv("data/personal_train.csv")

## Základný opis dát spolu s ich charakteristikami 

### Dataset other_train.csv

Pre predstavu tu je ukážka datasetu

In [4]:
df_other_train.head()

Unnamed: 0             name  \
0           0    Robert Plante   
1           1     Richard Huey   
2           2     Joseph Sable   
3           3  Gregory Preston   
4           4       Mike Riley   

                                             address                 race  \
0  3594 Johnson Green Suite 277\r\nJohnsonport, V...                White   
1             350 Lewis Haven\r\nAngelaton, IN 39486   Asian-Pac-Islander   
2  22509 Adrian Canyon Suite 799\r\nEast Timothyt...                White   
3           2674 Daryl Island\r\nLake Sean, TN 89363                White   
4  023 Joseph Estate Suite 799\r\nLake Andrewvill...                White   

        marital-status          occupation pregnant  education-num  \
0              Widowed       Other-service        f            5.0   
1        Never-married   Handlers-cleaners        f           13.0   
2        Never-married     Exec-managerial        f            9.0   
3   Married-civ-spouse     Exec-managerial        f           13.0   
4   Married-civ-spouse   Handlers-cleaners        f            9.0   

     relationship  capital-gain   education    fnlwgt  class  income  \
0       Unmarried           0.0         9th  220019.0    1.0   <=50K   
1   Not-in-family           0.0   Bachelors  196899.0    0.0   <=50K   
2       Unmarried           0.0     HS-grad  301591.0    0.0   <=50K   
3         Husband           0.0   Bachelors  310320.0    1.0    >50K   
4         Husband           0.0     HS-grad  237943.0    0.0   <=50K   

                                        medical_info  native-country  \
0  {'mean_glucose':'106.71875','std_glucose':'48....   United-States   
1  {'mean_glucose':'118.609375','std_glucose':'na...           Haiti   
2  {'mean_glucose':'129.4609375','std_glucose':'5...   United-States   
3  {'mean_glucose':'30.1015625','std_glucose':'37...   United-States   
4  {'mean_glucose':'124.0390625','std_glucose':'4...   United-States   

   hours-per-week  capital-loss workclass  
0             9.0           0.0   Private  
1            50.0           0.0   Private  
2            35.0           0.0   Private  
3            40.0           0.0   Private  
4            60.0           0.0   private

Pomocou tohto príkazu zistíme počet riadkov

In [5]:
df_other_train.shape[0]

3983

A teraz počet atribútov pre tento dataset

In [6]:
df_other_train.shape[1]

19

Typy atribútov zistíme týmto príkazom

In [7]:
df_other_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3983 entries, 0 to 3982
Data columns (total 19 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      3983 non-null   int64  
 1   name            3983 non-null   object 
 2   address         3983 non-null   object 
 3   race            3691 non-null   object 
 4   marital-status  3970 non-null   object 
 5   occupation      3961 non-null   object 
 6   pregnant        3969 non-null   object 
 7   education-num   3961 non-null   float64
 8   relationship    3954 non-null   object 
 9   capital-gain    3969 non-null   float64
 10  education       3972 non-null   object 
 11  fnlwgt          3968 non-null   float64
 12  class           3960 non-null   float64
 13  income          3972 non-null   object 
 14  medical_info    3966 non-null   object 
 15  native-country  3966 non-null   object 
 16  hours-per-week  3966 non-null   float64
 17  capital-loss    3973 non-null   f

Zistile sme, že tento dataset je zložený z atribútov:
- 6x float
- 1x int
- 12x object

Ďalej tu vidíme aj počet riadkov: 3983 ("RangeIndex: 3983 entries"),
a počet stĺpcov: 19 ("Data columns (total 19 columns)")

Nakoniec sa pozrieme na desktiptívne štatistiky.

In [8]:
df_other_train.describe()

Unnamed: 0  education-num  capital-gain         fnlwgt        class  \
count  3983.00000    3961.000000   3969.000000    3968.000000  3960.000000   
mean   1991.00000      10.052764   1134.530108  189963.996976     0.256313   
std    1149.93739       2.520019   8067.028170  102327.438172     0.436652   
min       0.00000       1.000000      0.000000   20308.000000     0.000000   
25%     995.50000       9.000000      0.000000  119475.250000     0.000000   
50%    1991.00000      10.000000      0.000000  180585.500000     0.000000   
75%    2986.50000      12.000000      0.000000  236173.000000     1.000000   
max    3982.00000      16.000000  99999.000000  795830.000000     1.000000   

       hours-per-week  capital-loss  
count     3966.000000   3973.000000  
mean        40.203984     72.926252  
std         12.317838    360.910859  
min          1.000000      0.000000  
25%         40.000000      0.000000  
50%         40.000000      0.000000  
75%         45.000000      0.000000  
max         99.000000   2824.000000

V tejto tabuľke vidíme ako sú: počet záznamov, priemer, štandardnú odchýlku, minimum, prvý kvartil, druhý kvartil (medián), tretí kvartil, maximum, pre všetky atribúty, ktoré obsahujú numerické hodnoty.

### Dataset personal_train.csv

Rovnako vykonáme opis dát aj pre druhý dataset

Ukážka datasetu

In [9]:
df_personal_train.head()

Unnamed: 0             name  \
0           0  Roscoe Bohannon   
1           1     Ernest Kline   
2           2  Harold Hendriks   
3           3   Randy Baptiste   
4           4  Anthony Colucci   

                                             address  age      sex  \
0  7183 Osborne Ways Apt. 651\r\nEast Andrew, OH ...   58     Male   
1  391 Ball Road Suite 961\r\nFlowersborough, IN ...   43   Female   
2     8702 Vincent Square\r\nNew Jerryfurt, CO 30614   51     Male   
3     2751 Harris Crossroad\r\nWest Ashley, CA 30311   66     Male   
4  904 Robert Cliffs Suite 186\r\nWest Kyle, CO 7...   45   Female   

  date_of_birth  
0    1961-04-27  
1    1976-09-12  
2    1967-11-12  
3    1953-06-02  
4    1974/09/11

Počet riadkov

In [10]:
df_personal_train.shape[0]

3933

Počet atribútov

In [11]:
df_personal_train.shape[1]

6

Typy atribútov

In [12]:
df_personal_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3933 entries, 0 to 3932
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Unnamed: 0     3933 non-null   int64 
 1   name           3933 non-null   object
 2   address        3933 non-null   object
 3   age            3933 non-null   int64 
 4   sex            3933 non-null   object
 5   date_of_birth  3933 non-null   object
dtypes: int64(2), object(4)
memory usage: 184.5+ KB


Zistile sme, že tento dataset je zložený z atribútov:
- 2x int
- 4x object

Ďalej tu vidíme aj počet riadkov: 3933 ("RangeIndex: 3933 entries"),
a počet stĺpcov: 6 ("Data columns (total 6 columns)")

A Nakoniec desktiptívne štatistiky.

In [13]:
df_personal_train.describe()

Unnamed: 0          age
count  3933.000000  3933.000000
mean   1966.000000    51.816934
std    1135.503633    11.814803
min       0.000000    -1.000000
25%     983.000000    44.000000
50%    1966.000000    52.000000
75%    2949.000000    60.000000
max    3932.000000   113.000000

## Párová analýza dát

## Identifikácia problémov v dátach spolu s predpokladaným scenárom riešenia v ďalšej fáze

Nevhodná štruktúra dát sa nachádza v atribúte medical_info v datasete other_train.csv, kde je atribút zložený z viacero zložiek (mean_glucose, std_glucose, kurtosis_glucose, skewness_glucose, mean_oxygen, std_oxygen, kurtosis_oxygen, skewness_oxygen), tento atribút chceme rozdeliť do viacero stĺpcov. Taktiež aj atribút address v datasete personal_train.csv obsahuje viac zložiek, ale nepredpokladáme, že by to ovplyvnilo naše riešenie, preto počítame s tým, že tento atribút nebudeme podrobnejšie deliť.

Teraz ideme zistiť či sa v datasete nenachádzajú duplicitné riadky.

In [66]:
df_other_train.duplicated().sum()

0

Duplicitné riadky sme nanašli, ale ešte nás zaujíma atribút name, pretože predpokladáme, že každý participant ma unikátne meno a nemáme viac záznamov daného človeka.

In [58]:
df_other_train.name.duplicated().sum()

50

Našli sme 50 duplicít, vypíšeme si pár z nich.

In [68]:
df_other_train[df_other_train.name.isin(df_other_train.name[df_other_train.name.duplicated()])].sort_values(by=['name']).head(6)

Unnamed: 0               name  \
2078        2078  Alex Stubblefield   
2682        2682  Alex Stubblefield   
2416        2416       Ashley Lyons   
467          467       Ashley Lyons   
2553        2553        Billy Clark   
3408        3408        Billy Clark   

                                                address    race  \
2078                 PSC 1476, Box 5156\r\nAPO AE 20184   White   
2682                 PSC 1476, Box 5156\r\nAPO AE 20184     NaN   
2416        4166 Sara Points\r\nHernandezview, WY 62853   White   
467         4166 Sara Points\r\nHernandezview, WY 62853   White   
2553  3859 Melissa Bridge Apt. 723\r\nShawnfurt, GA ...   White   
3408  3859 Melissa Bridge Apt. 723\r\nShawnfurt, GA ...   White   

           marital-status          occupation pregnant  education-num  \
2078   Married-civ-spouse     Protective-serv    FALSE           10.0   
2682   Married-civ-spouse                 NaN    FALSE           10.0   
2416   Married-civ-spouse   Handlers-cleaners        f            9.0   
467    Married-civ-spouse                 NaN        f            9.0   
2553             Divorced     Exec-managerial        f            9.0   
3408             Divorced     Exec-managerial        f            9.0   

        relationship  capital-gain      education    fnlwgt  class  income  \
2078         Husband           0.0   Some-college  194417.0    NaN    >50K   
2682         Husband           0.0   Some-college       NaN    1.0    >50K   
2416         Husband           NaN        HS-grad  155972.0    NaN    >50K   
467          Husband           0.0        HS-grad       NaN    0.0    >50K   
2553   Not-in-family           NaN        HS-grad  195516.0    0.0   <=50K   
3408   Not-in-family           0.0            NaN       NaN    0.0   <=50K   

                                           medical_info  native-country  \
2078  {'mean_glucose':'39.3046875','std_glucose':'42...   United-States   
2682  {'mean_glucose':'39.3046875','std_glucose':'42...   United-States   
2416                                                NaN   United-States   
467   {'mean_glucose':'111.1953125','std_glucose':'4...   United-States   
2553  {'mean_glucose':'103.8046875','std_glucose':'4...   United-States   
3408                                                NaN   United-States   

      hours-per-week  capital-loss   workclass  
2078            40.0           0.0   Local-gov  
2682            40.0           0.0   Local-gov  
2416            40.0           0.0     Private  
467              NaN           0.0     Private  
2553            40.0           0.0   Local-gov  
3408             NaN           0.0   Local-gov

Rovnaký postup zvolíme aj pre druhý dataset

In [63]:
df_personal_train.duplicated().sum()

0

In [64]:
df_personal_train.name.duplicated().sum()

0

Zistili sme, že v druhom datasete sa žiadne duplicity nenachádzajú.

Duplicitné záznamy chceme nahradiť len 1 záznamom a v prípade chýbajúcich hodnôt doplníme tieto hodnoty z druhého záznamu.

Nejednotné formáty dát sme našli v atribúte pregnant, kde to nadobúda hodnoty: FALSE, f, alebo TRUE, t, tak tieto hodnoty chceme zjednotiť. Ďalšie nejednotné hodnoty sme našli v atribúte date_of_birth, kde existuje viacero typov ("1976-08-29", "1974/09/11", "1955-01-04 00:00:00").

In [77]:
df_other_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3983 entries, 0 to 3982
Data columns (total 19 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      3983 non-null   int64  
 1   name            3983 non-null   object 
 2   address         3983 non-null   object 
 3   race            3691 non-null   object 
 4   marital-status  3970 non-null   object 
 5   occupation      3961 non-null   object 
 6   pregnant        3969 non-null   object 
 7   education-num   3961 non-null   float64
 8   relationship    3954 non-null   object 
 9   capital-gain    3969 non-null   float64
 10  education       3972 non-null   object 
 11  fnlwgt          3968 non-null   float64
 12  class           3960 non-null   float64
 13  income          3972 non-null   object 
 14  medical_info    3966 non-null   object 
 15  native-country  3966 non-null   object 
 16  hours-per-week  3966 non-null   float64
 17  capital-loss    3973 non-null   f

Ako vidíme, chýbajú nám hodnoty v atribútoch 3-18, tieto hodnoty chceme doplniť mediánom alebo priemerom.

In [76]:
df_personal_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3933 entries, 0 to 3932
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Unnamed: 0     3933 non-null   int64 
 1   name           3933 non-null   object
 2   address        3933 non-null   object
 3   age            3933 non-null   int64 
 4   sex            3933 non-null   object
 5   date_of_birth  3933 non-null   object
dtypes: int64(2), object(4)
memory usage: 184.5+ KB


V tomto datasete nechýbajú žiadne hodnoty.

Vychýlené hodnoty plánujeme nahradiť mediánom alebo priemerom.